pkg/notify/notifier/pushover/pushover.go

```diff
@@ -0,0 +1,228 @@
+	URL                = "https://api.pushover.net/1/messages.json"
+	DefaultTemplate    = `{{ template "nm.default.text" . }}`
+	PoMsgLimitAlert    = 99
+)
+
+type Notifier struct {
+	notifierCfg  *config.Config
+	pushover     []*config.Pushover
+	timeout      time.Duration
+	logger       log.Logger
+	template     *notifier.Template
+	templateName string
+	client       http.Client
+}
+
+type pushoverRequest struct {
+	pushoverMessage
+}
+
+type pushoverResponse struct {
+	Status  int      `json:"status"`
+	Request string   `json:"request"`
+	Errors  []string `json:"errors,omitempty"`
+	Receipt string   `json:"receipt,omitempty"`
+}
+
+func NewPushoverNotifier(logger log.Logger, receivers []config.Receiver, notifierCfg *config.Config) notifier.Notifier {
+
+	var path []string
+	opts := notifierCfg.ReceiverOpts
+	if opts != nil && opts.Global != nil {
+		path = opts.Global.TemplateFiles
+	}
+	tmpl, err := notifier.NewTemplate(path)
+	if err != nil {
+		_ = level.Error(logger).Log("msg", "PushoverNotifier: get template error", "error", err.Error())
+		return nil
+	}
+
+	n := &Notifier{
+		notifierCfg:  notifierCfg,
+		timeout:      DefaultSendTimeout,
+		logger:       logger,
+		template:     tmpl,
+		templateName: DefaultTemplate,
+		client:       http.Client{Timeout: DefaultSendTimeout},
+	}
+
+	if opts != nil && opts.Pushover != nil {
+
+		if opts.Pushover.NotificationTimeout != nil {
+			n.timeout = time.Second * time.Duration(*opts.Pushover.NotificationTimeout)
+			n.client.Timeout = n.timeout
+		}
+
+		if len(opts.Pushover.Template) > 0 {
+			n.templateName = opts.Pushover.Template
+		} else if opts.Global != nil && len(opts.Global.Template) > 0 {
+			n.templateName = opts.Global.Template
+		}
+	}
+
+	for _, r := range receivers {
+		receiver, ok := r.(*config.Pushover)
+		if !ok || receiver == nil {
+			continue
+		}
+
+		if receiver.PushoverConfig == nil {
+			_ = level.Warn(logger).Log("msg", "PushoverNotifier: ignore receiver because of empty config")
+			continue
+		}
+
+		n.pushover = append(n.pushover, receiver)
+	}
+
+	return n
+}
+
+func (n *Notifier) Notify(ctx context.Context, data template.Data) []error {
+
+	send := func(userKey string, c *config.Pushover) error {
+
+		start := time.Now()
+		defer func() {
+			_ = level.Debug(n.logger).Log("msg", "PushoverNotifier: send message", "userKey", userKey, "used", time.Since(start).String())
+		}()
+
+		// retrieve app's token
+		token, err := n.notifierCfg.GetCredential(c.PushoverConfig.Token)
+		if err != nil {
+			_ = level.Error(n.logger).Log("msg", "PushoverNotifier: get token secret", "userKey", userKey, "error", err.Error())
+			return err
+		}
+
+		// filter data by selector
+		filteredData := utils.FilterAlerts(data, c.Selector, n.logger)
+		if len(filteredData.Alerts) == 0 {
+			return nil
+		}
+
+		// split new data along with its Alerts to ensure each message is small enough to fit the Pushover's message length limit
+		splitData := utils.SplitAlerts(filteredData)
```

Comment:
Why do need to send notifications one by one? 
Depending on the template used, the size of each notification is different. You should use this [function](https://github.com/kubesphere/notification-manager/blob/master/pkg/notify/notifier/template.go#L121) to calculate the size of each notification to ensure that as many notifications as possible are sent each time. 

The purpose of this is to reduce the number of pushover API calls as much as possible to avoid the possible limit on the number of API calls.

Reply:
Resolved: https://github.com/txfs19260817/notification-manager/blob/0e75498c02eb381666cbb63da674b9714b0bfb74/pkg/notify/notifier/pushover/pushover.go#L127